### PR TITLE
Add Vapid support, mostly a subset of #11

### PR DESCRIPTION
**Pull request by wibblymat**
I was surprised to find the other day that we never resolved the VAPID PR (#11). 

Since the comment thread had stalled, I've put together this smaller counter-proposal that is hopefully less controversial.

@costinm @crhym3 @gauntface

### CODE REVIEW

File: webpush/push.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ import (
 
 const (
 	gcmURL     = "https://android.googleapis.com/gcm/send"
-	tempGcmURL = "https://gcm-http.googleapis.com/gcm"
+	tempGcmURL = "https://fcm-http.googleapis.com/fcm/send"
```
Comment (x1ddos) [inline webpush/push.go:28]: what if someone still has the old, gcm-http.googleapis.com/gcm, stored somewhere in their records?

File: webpush/push.go [hunk 2]
```diff
@@ -69,6 +69,25 @@ func NewPushRequest(sub *Subscription, message string, token string) (*http.Requ
 	return req, nil
 }
 
+// NewVapidRequest creates a valid Web Push HTTP request for sending a message
+// to a subscriber, using Vapid authentication. You can add more headers to
+// configure collapsing, TTL.
+func NewVapidRequest(sub *Subscription, message string, vapid *Vapid) (*http.Request, error) {
```
Comment (x1ddos) [inline webpush/push.go:75]: Would be really nice if I, as a developer, could understand when to use NewPushRequest vs NewVapidRequest.
The func name should probably be either NewVAPRequest or NewVAPIDRequest.

Comment (crhym3) [inline webpush/push.go]: Actually, no, I'd rather link the draft in vapid.go.

File: webpush/push.go [hunk 3]
```diff
@@ -69,6 +69,25 @@ func NewPushRequest(sub *Subscription, message string, token string) (*http.Requ
...
+	if err != nil {
+		return nil, err
+	}
+
+	tok, err := vapid.Token(sub.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Authorization", fmt.Sprintf(`Bearer %s`, tok))
+	req.Header.Set("Crypto-Key", req.Header.Get("Crypto-Key")+"; "+headerField("p256ecdsa", vapid.PublicKey()))
```
Comment (x1ddos) [inline webpush/push.go:86]: why "add" for authorization and "set" for crypto-key? I'd use "set" everywhere.

File: webpush/push.go [hunk 4]
```diff
@@ -89,5 +108,5 @@ func Send(client *http.Client, sub *Subscription, message, token string) (*http.
 
 // A helper for creating the value part of the HTTP encryption headers
 func headerField(headerType string, value []byte) string {
-	return fmt.Sprintf(`%s=%s`, headerType, base64.URLEncoding.EncodeToString(value))
+	return fmt.Sprintf(`%s=%s`, headerType, base64.RawURLEncoding.EncodeToString(value))
```
Comment (x1ddos) [inline webpush/push.go:111]: what does this have to do with vapid? :)
if you say it's required for vapid, I'll ask whether vapid-less will still work.

File: webpush/vapid.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	"encoding/base64"
+	"encoding/json"
+	"math/big"
+	"net/url"
+	"time"
+)
+
+var (
+	// JWT header is always the same, so use a pre-encoded string
+	vapidPrefix = []byte("eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9.")
```
Comment (x1ddos) [inline webpush/vapid.go:31]: what does it mean and where it's coming from.
I would add something like this to the doc comment:

```
// {"typ":"JWT","alg":"ES256"}
```

File: webpush/vapid.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+)
+
+type jwtBody struct {
+	Aud string `json:"aud"`
+	Sub string `json:"sub,omitempty"`
+	Exp int64  `json:"exp"`
+}
+
+// Vapid provides helpers for authenticating the sender of a push message
+type Vapid struct {
```
Comment (x1ddos) [inline webpush/vapid.go:42]: I'd reference the spec draft here. is it this one? https://tools.ietf.org/html/draft-ietf-webpush-vapid-01
Also, rename `Vapid` to `VAPID`.

File: webpush/vapid.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+	"encoding/json"
+	"math/big"
+	"net/url"
+	"time"
+)
+
+var (
+	// JWT header is always the same, so use a pre-encoded string
+	vapidPrefix = []byte("eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9.")
+	dot         = []byte(".")
```
Comment (x1ddos) [inline webpush/vapid.go:32]: dot is used in one place only and doesn't typically deserve to be a global var. inline where it's used.

File: webpush/vapid.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+// PublicKey returns an uncompressed EC point representing the
+// application's public key
+func (v *Vapid) PublicKey() []byte {
+	pub := v.priv.PublicKey
+	return elliptic.Marshal(pub.Curve, pub.X, pub.Y)
+}
+
+// Token creates a token with the specified endpoint, using configured Sub id
+// and a default expiration (1h).
+func (v *Vapid) Token(aud string) (string, error) {
```
Comment (x1ddos) [inline webpush/vapid.go:59]: "... with the specified endpoint **aud**, using configured **subscription ID** ..."
expiration is hard-coded so it is not a default. just say something like "Token is valid for the next hour".

File: webpush/vapid.go [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+	t, err := json.Marshal(jwt)
+	if err != nil {
+		return "", err
+	}
+	enc := base64.RawURLEncoding
+
+	t64 := make([]byte, enc.EncodedLen(len(t)))
+	enc.Encode(t64, t)
+
+	token := make([]byte, len(t)+len(vapidPrefix)+100)
```
Comment (x1ddos) [inline webpush/vapid.go:78]: why 100?

File: webpush/vapid.go [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+	token = append(token[:0], vapidPrefix...)
+	token = append(token, t64...)
+
+	hasher := crypto.SHA256.New()
+	hasher.Write(token)
+
+	r, s, err := ecdsa.Sign(rand.Reader, v.priv, hasher.Sum(nil))
```
Comment (x1ddos) [inline webpush/vapid.go:85]: just use sha256.Sum:

``` go
h := sha256.Sum(token)
sigr, sigs, err := ecdsa.Sign(rand.Reader, v.priv, h[:])
```

File: webpush/vapid.go [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+	if err != nil {
+		return "", err
+	}
+	// Vapid key is 32 bytes
+	keyBytes := 32
```
Comment (x1ddos) [inline webpush/vapid.go:90]: add a comment why it is 32. also `keyBytes` is not the best name as it normally implies a slice of bytes. use something like `length` or `size` - it's short and clear.

Comment (x1ddos) [inline webpush/vapid.go:90]: also, I would probably make it `64`.

File: webpush/vapid.go [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+	sig := make([]byte, 2*keyBytes)
+
+	rBytes := r.Bytes()
+	copy(sig[keyBytes-len(rBytes):], rBytes)
+
+	sBytes := s.Bytes()
+	copy(sig[2*keyBytes-len(sBytes):], sBytes)
+
+	sigB64 := make([]byte, enc.EncodedLen(len(sig)))
+	enc.Encode(sigB64, sig)
```
Comment (x1ddos) [inline webpush/vapid.go:100]: hard to read because of so many cameCases and empty lines. here's an alternative:

``` go
const size = 64

sig := make([]byte, size)
copy(sig[size/2-len(sigr.Bytes()):], sigr.Bytes())
copy(sig[size-len(sigs.Bytes()):], sigs.Bytes())

sigb64 = make([]byte, enc.EncodedLen(len(sig))
enc.Encode(sigb64, sig)
```

File: webpush/vapid.go (hunk 5 above)
Comment (x1ddos) [inline webpush/vapid.go:31]: also, remove the trailing dot from here. it makes Vapid's Token method harder to understand.

File: webpush/vapid.go [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+
+	token = append(token, dot...)
+	token = append(token, sigB64...)
+	return string(token), nil
```
Comment (x1ddos) [inline webpush/vapid.go:104]: you could use bytes.Buffer for token:

``` go
var token bytes.Buffer
token.Write(vapidPrefix)
token.WriteRune('.')
token.Write(t64)
...
token.WriteRune('.')
token.Write(sigb64)
return token.String(), nil
```

File: webpush/vapid.go [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+type jwtBody struct {
+	Aud string `json:"aud"`
+	Sub string `json:"sub,omitempty"`
+	Exp int64  `json:"exp"`
+}
+
+// Vapid provides helpers for authenticating the sender of a push message
+type Vapid struct {
+	// The application's private key
+	priv *ecdsa.PrivateKey
```
Comment (x1ddos) [inline webpush/vapid.go:44]: why not have it exposed? that way one can also create a VAPID as simple as:

``` go
v := VAPID{Sub: s, Key: k}
```

assuming I'm in possession of the `k`.

File: webpush/vapid.go [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+
+	token = append(token, dot...)
+	token = append(token, sigB64...)
+	return string(token), nil
+}
+
+// NewVapid constructs a new Vapid generator from a p256 EC private key
+func NewVapid(privateKey []byte) *Vapid {
+	d := new(big.Int).SetBytes(privateKey)
+	x, y := curve.ScalarMult(curve.Params().Gx, curve.Params().Gy, privateKey)
```
Comment (x1ddos) [inline webpush/vapid.go:110]: I've been searching where curve is defined all over the place. Please, just use `elliptic.P256()` and remove `curve` global var from encrypt.go (remove it in another PR).

File: webpush/vapid.go [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+
+	sigB64 := make([]byte, enc.EncodedLen(len(sig)))
+	enc.Encode(sigB64, sig)
+
+	token = append(token, dot...)
+	token = append(token, sigB64...)
+	return string(token), nil
+}
+
+// NewVapid constructs a new Vapid generator from a p256 EC private key
```
Comment (x1ddos) [inline webpush/vapid.go:107]: "... from a P-256 ..."
also, what's generator?

**Comment by x1ddos** [comment]
Reviewed.

**Comment by x1ddos** [comment]
Forgot to ask: tests?

File: webpush/vapid.go (hunk 14 above)
Comment (x1ddos) [inline webpush/vapid.go:44]: if you do the above (expose VAPID.key), you could go even further and define it as follows:

``` go
type VAPID ecdsa.PrivateKey

func (vid *VAPID) Token(sub, aud string) (string, error) {
  ...
}
```

File: webpush/push.go (hunk 2 above)
Comment (x1ddos) [inline webpush/push.go:75]: Also, what's the equivalent of `Send` for `NewVAPIDRequest`?

File: webpush/vapid.go (hunk 14 above)
Comment (x1ddos) [inline webpush/vapid.go:44]: Alternatively, what prevents you from having something like this?

``` go
func VAPToken(key *ecdsa.PrivateKey, sub, aud string) (string, error) {
  ...
}
```

File: webpush/vapid.go (hunk 10 above)
Comment (x1ddos) [reply webpush/vapid.go:85]: I don't think you can actually use sha256.Sum directly. Please, ignore the above comment.

File: webpush/vapid.go [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+	enc := base64.RawURLEncoding
+
+	t64 := make([]byte, enc.EncodedLen(len(t)))
+	enc.Encode(t64, t)
+
+	token := make([]byte, len(t)+len(vapidPrefix)+100)
+	token = append(token[:0], vapidPrefix...)
+	token = append(token, t64...)
+
+	hasher := crypto.SHA256.New()
```
Comment (x1ddos) [inline webpush/vapid.go:82]: import "crypto/sha256"
sha256.New()

File: webpush/vapid.go (hunk 12 above)
Comment (x1ddos) [reply webpush/vapid.go:100]: I've actually done a very similar thing recently, if it helps:
https://github.com/golang/crypto/blob/b35ccbc95a0eaae49fb65c5d627cb7149ed8d1ab/acme/internal/acme/jws.go#L107-L119

**Comment by davinche** [comment]
Has this been tested? Does this implementation work with chrome? I also added support for VAPID on top of this library, but it is only working in firefox. On chrome, FCM is sending back a 400 "UnauthorizedRegistration" error.

File: webpush/vapid.go [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+func (v *Vapid) Token(aud string) (string, error) {
+	url, err := url.Parse(aud)
+	if err != nil {
+		return "", err
+	}
+	jwt := jwtBody{Aud: url.Scheme + "://" + url.Host}
+	if v.Sub != "" {
+		jwt.Sub = v.Sub
+	}
+	jwt.Exp = int64(time.Now().Unix() + 3600)
```
Comment (davinche) [inline webpush/vapid.go:68]: Use `time.Add()` to create the new expiration. Eg: `time.Now().Add(time.Hour * 12).Unix()`

**Comment by ofavre** [comment]
I've just tested this with Firefox 50.0.2 and Chrome 55.0.2883.75, it works fine!

Will this be merged someday?
Is this library defunct because of the raise of the [web-push](https://github.com/web-push-libs/web-push) Node.JS lib?

@davinche:
You get an `UnauthorizedRegistration` 400 response if the manifest contains a `gcm_sender_id` key for Chrome and I use `applicationServerKey` when subscribing and `NewVapidRequest()` from this pull request.
I guess this comes from the fact that an `Authorization: key=XYZ` is expected when using a GCM SenderID and this header is used with a JWT when using VAPID. That and the end-point that does not mention the subscription id.

File: webpush/push.go [hunk 19]
```diff
@@ -69,6 +69,25 @@ func NewPushRequest(sub *Subscription, message string, token string) (*http.Requ
...
+	req, err := NewPushRequest(sub, message, "")
+	if err != nil {
+		return nil, err
+	}
+
+	tok, err := vapid.Token(sub.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Authorization", fmt.Sprintf(`Bearer %s`, tok))
```
Comment (ofavre) [inline webpush/push.go:85]: The [latest spec](https://tools.ietf.org/html/draft-ietf-webpush-vapid-01) switched from `Bearer` to `WebPush`. According to my test, this is now supported by both FCM's and Mozilla's servers.

**Comment by davinche** [comment]
@ofavre The UnauthorizedRegistration error I was getting was actually caused by the JWT Header serialization. More details can be found in my reply to this stackoverflow post: http://stackoverflow.com/questions/39336523/webpushvapid-request-fails-with-400-unauthorizedregistration

The gist of it all is that json/encoding libary in Go produced a marshalled JWT header that is different than what chrome expected. Apparently order of the fields mattered ¯\_(ツ)_/¯. I think this is a bug in chrome.

This PR avoids it all by including a static string: <https://github.com/GoogleChrome/push-encryption-go/blob/d00644ee80b2c526918c0a6a88c63d4206fccb0a/webpush/vapid.go#L31>

**Comment by avinassh** [comment]
Any update on this?

File: webpush/push.go (hunk 4 above)
Comment (anaskhan96) [reply webpush/push.go:111]: `URLEncoding` and `RawURLEncoding` give padded and non-padded `base64url` encoded strings respectively. Vapid apparently accepts only non-padded `base64url` strings.

**Comment by anaskhan96** [comment]
@avinassh I have been waiting for the vapid support too. In the meantime, [this](https://github.com/SherClockHolmes/webpush-go) library has been helping me out.

File: webpush/push.go (hunk 1 above)
Comment (gauntface) [reply webpush/push.go:28]: Where is this used? It shouldn't be here at all. I think a long time ago this replacement was needed, but not anymore.

File: webpush/push.go (hunk 2 above)
Comment (gauntface) [reply webpush/push.go:75]: The naming here feels really awkward and it may just be me not following Go convention, but "NewVapidRequest" really doesn't mean anything to me (i.e. does this just make the request, or is it just a new set of VAPID headers or is it actually sending a push message using Vapid?) 

@x1ddos what was your thinking behind "NewVAPRequest" vs "NewVAPIDRequest"?

File: webpush/push.go (hunk 19 above)
Comment (gauntface) [reply webpush/push.go:85]: +1 This change.

File: webpush/vapid.go (hunk 5 above)
Comment (gauntface) [reply webpush/vapid.go:31]: I would strongly recommend not just hardcoding this kind of this - makes it extremely hard to parse / reason with and should anything change in the future, it will be unnecessarily difficult to just make the change.

File: webpush/vapid.go (hunk 7 above)
Comment (gauntface) [reply webpush/vapid.go:32]: This is again very very weird to see.

File: webpush/vapid.go (hunk 11 above)
Comment (gauntface) [reply webpush/vapid.go:90]: @x1ddos why make it 64?

Comment (x1ddos) [reply webpush/vapid.go:90]: @gauntface see the next comment, suggesting an alternative, the snippet which starts with `const size = 64 ...`.

File: webpush/push.go (hunk 4 above)
Comment (x1ddos) [reply webpush/push.go:111]: @anaskhan96 ok so my question remains: with this change, will it still work when sending msgs without VAP ID?

File: webpush/push.go (hunk 2 above)
Comment (x1ddos) [reply webpush/push.go:75]: @gauntface merely Go idiomatic naming convention: https://github.com/golang/go/wiki/CodeReviewComments#initialisms
But the other issue remains: if I haven't seen the package before, I probably wouldn't know whether I should use `NewPushRequest` or the `NewVAPRequest` and when.

I don't know what the usage is like of this package, but maybe there's still time to just make a breaking change now, have one function which works for both, with and without VAP.

File: webpush/push.go (hunk 1 above)
Comment (x1ddos) [reply webpush/push.go:28]: Would be nice to know what the usage is like of this package. If it's low, then this can probably be removed.

**Comment by x1ddos** [comment]
@gauntface I'll take this PR and apply all suggestions. Will ping for a review.

**Comment by gauntface** [comment]
@x1ddos thank you Alex. I'll get to work on a PR for integration tests (expect awful Go)

@avinassh @anaskhan96 apologies for the delay on all of this. I've inherited the project and once we have tests and VAPID set up I'll be moving this over to web-push-libs org so it can get wider community eyes and support.

**Comment by nicosavini** [comment]
Hi any updates on this? Thanks,

**Comment by gauntface** [comment]
I would recommend using: https://github.com/SherClockHolmes/webpush-go

This module had a number of quirks / issues that feels wasted to develop this module when the other seems further ahead in it's implementation

**Comment by gauntface** [comment]
Going to deprecate this project in favor of https://github.com/SherClockHolmes/webpush-go
